builder/main.py

```diff
@@ -109,6 +115,100 @@ def generate_uf2(target, source, env):
+    if len(res) == 0:
+        sys.stderr.write(
+            "Error: Could not parse filesystem size expression '%s'."
+            " Will treat as size = 0.\n" % str(expression))
+        return 0
+    # access first result
+    number, factor = res[0]
+    number = float(number)
+    number *= conversion_factors[factor.upper()]
+    return int(number)
+
+def fetch_fs_size(env):
+    # follow generation formulas from makeboards.py for Earle Philhower core
+    # given the total flash size, a user can specify
+    # the amount for the filesystem (0MB, 2MB, 4MB, 8MB, 16MB)
+    # via board_build.filesystem_size,
+    # and we will calculate the flash size and eeprom size from that.
+    flash_size = board.get("upload.maximum_size")
+    filesystem_size = board.get("build.filesystem_size", "0MB")
+    filesystem_size_int = convert_size_expression_to_int(filesystem_size)
+
+    maximum_sketch_size = flash_size - 4096 - filesystem_size_int
+
+    print("Flash size: %.2fMB" % (flash_size / 1024.0 / 1024.0))
+    print("Sketch size: %.2fMB" % (maximum_sketch_size / 1024.0 / 1024.0))
+    print("Filesystem size: %.2fMB" % (filesystem_size_int / 1024.0 / 1024.0))
+
+    eeprom_start = 0x10000000 + flash_size - 4096
+    fs_start = 0x10000000 + flash_size - 4096 - filesystem_size_int
+    fs_end = 0x10000000 + flash_size - 4096
+
+    if maximum_sketch_size <= 0:
+        sys.stderr.write(
+            "Error: Filesystem too large for given flash. "
+            "Can at max be flash size - 4096 bytes. "
+            "Available sketch size with current "
+            "config would be %d bytes.\n" % maximum_sketch_size)
+        sys.stderr.flush()
+        env.Exit(1)
+
+    env["PICO_FLASH_LENGTH"] = maximum_sketch_size
+    env["PICO_EEPROM_START"] = eeprom_start
+    env["FS_START"] = fs_start
+    env["FS_END"] = fs_end
+    # LittleFS configuration paramters taken from
+    # https://github.com/earlephilhower/arduino-pico-littlefs-plugin/blob/master/src/PicoLittleFS.java
+    env["FS_PAGE"] = 256
+    env["FS_BLOCK"] = 4096
+
+    print("Maximium Sketch size: %d "
+        "EEPROM start: %s Filesystem start: %s "
+        "Filesystem end: %s" % 
+        (maximum_sketch_size, hex(eeprom_start), hex(fs_start), hex(fs_end)))
+
+
+def __fetch_fs_size(target, source, env):
+    fetch_fs_size(env)
+    return (target, source)
+
+env.Append(
+    BUILDERS=dict(
+        DataToBin=Builder(
+            action=env.VerboseAction(" ".join([
+                '"$MKFSTOOL"',
+                "-c", "$SOURCES",
+                "-p", "$FS_PAGE",
+                "-b", "$FS_BLOCK",
+                "-s", "${FS_END - FS_START}",
+                "$TARGET"
+            ]), "Building file system image from '$SOURCES' directory to $TARGET"),
+            emitter=__fetch_fs_size,
+            source_factory=env.Dir,
+            suffix=".bin"
+        )
+    )
+)
+
+# store function to get infno about filesystems for builder scripts.
+env["fetch_fs_size"] = fetch_fs_size
```

Comment:
May it be a good idea to rename the function to `__fetch_fs_size` to avoid polluting the env? Also, would be great to move these helper functions to the top of the file, something like in the [`espressif32` platform](https://github.com/platformio/platform-espressif32/blob/develop/builder/main.py).

Note: `infno` typo.

Reply:
That is actually a tricky one, I "save" that function in the `env` here to be able to use the same functionality later in the builder script at https://github.com/earlephilhower/arduino-pico/blob/master/tools/platformio-build.py#L224-L230 to avoid code duplication. Is there a better way to share them?

Reply:
What you did here is just fine, it just would be great to make it a bit tidier as in espressif32.

Reply:
Function in `env` was renamed to `__fetch_fs_size`, builder script on Arduino-Pico was adapted accordingly.